src/ipc/ipc4/helper.c

```diff
@@ -453,32 +459,85 @@ int ipc_comp_connect(struct ipc *ipc, ipc_pipe_comp_connect *_connect)
 	if (!cpu_is_me(source->ipc_config.core) && !cross_core_bind)
 		return ipc4_process_on_core(source->ipc_config.core, false);
 
-	/* these might call comp_ipc4_get_attribute_remote() if necessary */
-	ret = comp_get_attribute(source, COMP_ATTR_BASE_CONFIG, &source_src_cfg);
-	if (ret < 0) {
-		tr_err(&ipc_tr, "failed to get base config for module %#x", dev_comp_id(source));
-		return IPC4_FAILURE;
+	size = MODULE_MAX_SINKS * sizeof(struct ipc4_output_pin_format) +
+		MODULE_MAX_SOURCES * sizeof(struct ipc4_input_pin_format);
+
+	/* get obs from the base config extension if the src queue ID is non-zero */
+	if (bu->extension.r.src_queue) {
+		struct ipc4_output_pin_format out_fmt;
+		size_t output_fmt_offset;
+
+		src_basecfg_ext = rzalloc(SOF_MEM_ZONE_RUNTIME, 0, SOF_MEM_CAPS_RAM,
+					  sizeof(*src_basecfg_ext) + size);
+		if (!src_basecfg_ext)
+			return IPC4_OUT_OF_MEMORY;
+
+		ret = comp_get_attribute(source, COMP_ATTR_BASE_CONFIG_EXT, src_basecfg_ext);
+		if (!ret) {
+			size_t in_fmt_size = sizeof(struct ipc4_input_pin_format);
+			size_t out_fmt_size = sizeof(struct ipc4_output_pin_format);
+			output_fmt_offset =
+				src_basecfg_ext->nb_input_pins * in_fmt_size +
+				bu->extension.r.src_queue * out_fmt_size;
+
+			memcpy_s(&out_fmt, sizeof(struct ipc4_output_pin_format),
+				 &src_basecfg_ext->pin_formats[output_fmt_offset],
+				 sizeof(struct ipc4_output_pin_format));
+			obs = out_fmt.obs;
+		}
+		rfree(src_basecfg_ext);
 	}
 
-	ret = comp_get_attribute(sink, COMP_ATTR_BASE_CONFIG, &sink_src_cfg);
-	if (ret < 0) {
-		tr_err(&ipc_tr, "failed to get base config for module %#x", dev_comp_id(sink));
-		return IPC4_FAILURE;
+	/* get obs from base config if src queue ID is 0 or if base config extn is missing */
+	if (!obs) {
+		/* these might call comp_ipc4_get_attribute_remote() if necessary */
+		ret = comp_get_attribute(source, COMP_ATTR_BASE_CONFIG, &source_src_cfg);
+
+		if (ret < 0) {
+			tr_err(&ipc_tr, "failed to get base config for src module %#x",
+			       dev_comp_id(source));
+			return IPC4_FAILURE;
+		}
+		obs = source_src_cfg.obs;
 	}
 
-	/* create a buffer
-	 * in case of LL -> LL or LL->DP
-	 *	size = 2*obs of source module (obs is single buffer size)
-	 * in case of DP -> LL
-	 *	size = 2*ibs of destination (LL) module. DP queue will handle obs of DP module
-	 */
-	uint32_t buf_size;
+	/* get ibs from the base config extension if the sink queue ID is non-zero */
+	if (bu->extension.r.dst_queue) {
+		struct ipc4_input_pin_format in_fmt;
+		size_t input_fmt_offset;
+
+		sink_basecfg_ext = rzalloc(SOF_MEM_ZONE_RUNTIME, 0, SOF_MEM_CAPS_RAM,
```

Comment:
Btw, if these are always copied and freed (in the same func), why not just have on the stack ?

Reply:
because basecfg_ext structure contains a flexible array member for the number of pin formats

Reply:
could we add a component method to get a pin format?